crates/core/app/src/app_version/component.rs

```diff
@@ -0,0 +1,138 @@
+use std::fmt::Write as _;
+
+use anyhow::{anyhow, Context};
+use cnidarium::{StateDelta, StateRead, StateWrite, Storage};
+
+use super::APP_VERSION;
+
+fn version_to_software_version(version: u64) -> &'static str {
+    match version {
+        1 => "v0.70.x",
+        2 => "v0.73.x",
+        3 => "v0.74.x",
+        4 => "v0.75.x",
+        5 => "v0.76.x",
+        6 => "v0.77.x",
+        7 => "v0.79.x",
+        8 => "v0.80.x",
+        _ => "unknown",
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum CheckContext {
+    Running,
+    Migration,
+}
+
+fn check_version(ctx: CheckContext, expected: u64, found: Option<u64>) -> anyhow::Result<()> {
+    let found = match (expected, found) {
+        (x, Some(y)) if x != y => y,
+        _ => return Ok(()),
+    };
+    match ctx {
+        CheckContext::Running => {
+            let expected_name = version_to_software_version(expected);
+            let found_name = version_to_software_version(expected);
+            let mut error = String::new();
+            error.push_str("app version mismatch:\n");
+            write!(
+                &mut error,
+                "  expected {} (penumbra {})\n",
+                expected, expected_name
+            )?;
+            write!(&mut error, "  found {} (penumbra {})\n", found, found_name)?;
+            write!(
+                &mut error,
+                "make sure you're running penumbra {}",
+                expected_name
+            )?;
+            Err(anyhow!(error))
+        }
+        CheckContext::Migration => {
+            let expected_name = version_to_software_version(expected);
+            let found_name = version_to_software_version(expected);
+            let mut error = String::new();
+            error.push_str("app version mismatch:\n");
+            write!(
+                &mut error,
+                "  expected {} (penumbra {})\n",
+                expected, expected_name
+            )?;
+            write!(&mut error, "  found {} (penumbra {})\n", found, found_name)?;
+            write!(
+                &mut error,
+                "this migration should be run with penumbra {} instead",
+                version_to_software_version(expected + 1)
+            )?;
+            Err(anyhow!(error))
+        }
+    }
+}
+
+fn state_key() -> Vec<u8> {
+    b"penumbra_app_version_safeguard".to_vec()
+}
+
+async fn read_app_version_safeguard<S: StateRead>(s: &S) -> anyhow::Result<Option<u64>> {
+    const CTX: &'static str = "while reading app_version_safeguard";
+
+    let res = s.nonverifiable_get_raw(&state_key()).await.context(CTX)?;
+    match res {
+        None => Ok(None),
+        Some(x) => {
+            let bytes: [u8; 8] = x
+                .try_into()
+                .map_err(|bad: Vec<u8>| {
+                    anyhow!("expected bytes to have length 8, found: {}", bad.len())
+                })
+                .context(CTX)?;
+            Ok(Some(u64::from_le_bytes(bytes)))
```

Comment:
https://github.com/penumbra-zone/penumbra/blob/main/crates/core/component/governance/src/component/view.rs#L978

Looking at this, is there an example of reading?

Because you can indeed write a Vec<u8> using write_proto, but you still need to choose how to encode the u64 that way, unless we specifically made a bespoke proto type for this

Reply:
https://rustdoc.penumbra.zone/main/penumbra_proto/trait.Message.html#impl-Message-for-u64 oh I see, it's because u64 implements Message